README.md
# googlesheets

Download a table from Google Sheets.

# Development

First, set up your enviroment. It's just Docker.

Run `docker build .` to test that everything works.

Then do the Development Dance:

1. Write a failing test to `tests/`
2. Run `docker build .` to test it fails
3. Fix the code so all tests pass
4. Submit a pull request

# Deployment

1. Edit the version in `pyproject.toml`
2. Write an entry to `CHANGELOG.md`
3. Run `docker build .` one last time
4. `git push`


CHANGELOG.md
2021-01-28.01
-------------

* Move to a new GitHub repository.
* Support non-English error messages


tests/test_fetch.py
import contextlib
import tempfile
from pathlib import Path
from typing import Any, ContextManager, Dict, Optional

import httpx
from cjwmodule.http import httpfile
from cjwmodule.testing.i18n import cjwmodule_i18n_message, i18n_message
from pytest_httpx import HTTPXMock

from googlesheets import FetchResult, fetch_arrow

DEFAULT_SECRET = {
    "name": "x",
    "secret": {
        # As returned by fetcher.secrets.process_secret_oauth2()
        "token_type": "Bearer",
        "access_token": "an-access-token",
    },
}
DEFAULT_FILE = {
    "id": "aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj",
    "name": "Police Data",
    "url": "http://example.org/police-data",
    "mimeType": "application/vnd.google-apps.spreadsheet",
}


def P(file=DEFAULT_FILE, has_header=True):
    return dict(file=file, has_header=has_header)


def secrets(secret: Optional[Dict[str, Any]]) -> Dict[str, Any]:
    if secret:
        return {"google_credentials": secret}
    else:
        return {}


@contextlib.contextmanager
def _fetch(
    params: Dict[str, Any], secrets: Dict[str, Any]
) -> ContextManager[FetchResult]:
    with tempfile.NamedTemporaryFile() as tf:
        output_path = Path(tf.name)
        yield fetch_arrow(params, secrets, None, None, output_path)


def test_fetch_nothing():
    with _fetch(P(file=None), {}) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [i18n_message("error.params.noFile")]


def test_fetch_native_sheet(httpx_mock: HTTPXMock):
    body = b"A,B\nx,y\nz,a"
    httpx_mock.add_response(
        url="https://www.googleapis.com/drive/v3/files/aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj/export?mimeType=text%2Fcsv",
        data=body,
        headers={"Content-Type": "text/csv", "Content-Length": str(len(body))},
    )
    with _fetch(P(), secrets(DEFAULT_SECRET)) as (path, errors):
        assert errors == []
        with httpfile.read(path) as (_, __, headers, body_path):
            assert body_path.read_bytes() == body
            assert headers == [("Content-Type", "text/csv"), ("Content-Length", "11")]


def test_fetch_csv_file(httpx_mock: HTTPXMock):
    body = b"A,B\nx,y\nz,a"
    httpx_mock.add_response(
        url="https://www.googleapis.com/drive/v3/files/aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj?alt=media",
        data=body,
        headers={"Content-Type": "text/csv", "Content-Length": str(len(body))},
    )
    with _fetch(
        P(file={**DEFAULT_FILE, "mimeType": "text/csv"}), secrets(DEFAULT_SECRET)
    ) as (path, errors):
        assert errors == []
        with httpfile.read(path) as (_, __, headers, body_path):
            assert body_path.read_bytes() == body
            assert headers == [("Content-Type", "text/csv"), ("Content-Length", "11")]


def test_fetch_tsv_file(httpx_mock: HTTPXMock):
    body = b"A\tB\nx\ty\nz\ta"
    httpx_mock.add_response(
        url="https://www.googleapis.com/drive/v3/files/aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj?alt=media",
        data=body,
        headers={
            "Content-Type": "text/tab-separated-values",
            "Content-Length": str(len(body)),
        },
    )
    with _fetch(
        P(file={**DEFAULT_FILE, "mimeType": "text/tab-separated-values"}),
        secrets(DEFAULT_SECRET),
    ) as (path, errors):
        assert errors == []
        with httpfile.read(path) as (_, __, headers, body_path):
            assert body_path.read_bytes() == body
            assert headers == [
                ("Content-Type", "text/tab-separated-values"),
                ("Content-Length", "11"),
            ]


def test_fetch_xls_file(httpx_mock: HTTPXMock):
    body = b"abcd"
    httpx_mock.add_response(
        url="https://www.googleapis.com/drive/v3/files/aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj?alt=media",
        data=body,
        headers={
            "Content-Type": "application/vnd.ms-excel",
            "Content-Length": str(len(body)),
        },
    )
    with _fetch(
        P(file={**DEFAULT_FILE, "mimeType": "application/vnd.ms-excel"}),
        secrets(DEFAULT_SECRET),
    ) as (path, errors):
        assert errors == []
        with httpfile.read(path) as (_, __, headers, body_path):
            assert body_path.read_bytes() == body
            assert headers == [
                ("Content-Type", "application/vnd.ms-excel"),
                ("Content-Length", "4"),
            ]


def test_fetch_xlsx_file(httpx_mock: HTTPXMock):
    body = b"abcd"
    httpx_mock.add_response(
        url="https://www.googleapis.com/drive/v3/files/aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj?alt=media",
        data=body,
        headers={
            "Content-Type": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            "Content-Length": str(len(body)),
        },
    )
    with _fetch(
        P(
            file={
                **DEFAULT_FILE,
                "mimeType": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
            }
        ),
        secrets(DEFAULT_SECRET),
    ) as (path, errors):
        assert errors == []
        with httpfile.read(path) as (_, __, headers, body_path):
            assert body_path.read_bytes() == body
            assert headers == [
                (
                    "Content-Type",
                    "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
                ),
                ("Content-Length", "4"),
            ]


def test_missing_secret_error():
    with _fetch(P(), {}) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [
            i18n_message("error.secrets.noCredentials")
        ]


def test_http_401_unauthorized(httpx_mock: HTTPXMock):
    httpx_mock.add_response(status_code=401)
    with _fetch(P(), secrets=secrets(DEFAULT_SECRET)) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [i18n_message("error.http.status401")]


def test_http_403_forbidden(httpx_mock: HTTPXMock):
    httpx_mock.add_response(status_code=403)
    with _fetch(P(), secrets=secrets(DEFAULT_SECRET)) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [i18n_message("error.http.status403")]


def test_http_404_not_found(httpx_mock: HTTPXMock):
    httpx_mock.add_response(status_code=404)
    with _fetch(P(), secrets=secrets(DEFAULT_SECRET)) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [i18n_message("error.http.status404")]


def test_http_not_200_ok_generic(httpx_mock: HTTPXMock):
    httpx_mock.add_response(status_code=429)
    with _fetch(P(), secrets=secrets(DEFAULT_SECRET)) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [
            cjwmodule_i18n_message(
                "http.errors.HttpErrorNotSuccess",
                {"status_code": 429, "reason": "Too Many Requests"},
            )
        ]


def test_unhandled_http_error(httpx_mock: HTTPXMock):
    def raise_timeout(request, ext):
        raise httpx.ReadTimeout("Gave up", request=request)

    httpx_mock.add_callback(raise_timeout)
    with _fetch(P(), secrets=secrets(DEFAULT_SECRET)) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [
            # googlesheet should pass through cjwmodule.http.client's message
            cjwmodule_i18n_message("http.errors.HttpErrorTimeout")
        ]


def test_secret_error():
    # Workbench runs oauth _before fetch_. It passes the error to fetch().
    with _fetch(
        P(),
        secrets=secrets(
            {
                "name": "x",
                "error": {
                    "id": "py.fetcher.secrets._refresh_oauth2_token.error.general",
                    "arguments": {
                        "status_code": 400,
                        "error": "invalid_grant",
                        "description": "Token has been expired or revoked.",
                    },
                    "source": None,
                },
            }
        ),
    ) as (path, errors):
        assert path.read_bytes() == b""
        assert [e.message for e in errors] == [
            (
                "py.fetcher.secrets._refresh_oauth2_token.error.general",
                {
                    "status_code": 400,
                    "error": "invalid_grant",
                    "description": "Token has been expired or revoked.",
                },
                None,
            )
        ]


tests/test_render.py
import contextlib
import io
import tempfile
from pathlib import Path
from typing import Any, ContextManager, Dict, List, Optional, Tuple

import pyarrow as pa
from cjwmodule.http import httpfile
from cjwmodule.testing.i18n import i18n_message

from googlesheets import FetchResult, RenderError, render

DEFAULT_FILE = {
    "id": "aushwyhtbndh7365YHALsdfsdf987IBHJB98uc9uisdj",
    "name": "Police Data",
    "url": "http://example.org/police-data",
    "mimeType": "application/vnd.google-apps.spreadsheet",
}


def P(file=DEFAULT_FILE, has_header=True):
    return dict(file=file, has_header=has_header)


@contextlib.contextmanager
def _temp_parquet_file(table: pa.Table) -> ContextManager[Path]:
    with tempfile.NamedTemporaryFile() as tf:
        path = Path(tf.name)
        pa.parquet.write_table(table, path, version="2.0", compression="SNAPPY")
        yield path


@contextlib.contextmanager
def _temp_httpfile(
    url: str,
    status_line: str,
    body: bytes,
    headers: List[Tuple[str, str]] = [("Content-Type", "text/html; charset=utf-8")],
) -> ContextManager[Path]:
    with tempfile.NamedTemporaryFile() as tf:
        path = Path(tf.name)
        httpfile.write(path, {"url": url}, status_line, headers, io.BytesIO(body))
        yield path


def _assert_table_file(path: Path, expected: Optional[pa.Table]) -> None:
    if expected is None:
        assert path.stat().st_size == 0
        return
    else:
        assert path.stat().st_size > 0

    with pa.ipc.open_file(path) as f:
        actual = f.read_all()
    assert actual.column_names == expected.column_names
    for actual_column, expected_column in zip(
        actual.itercolumns(), expected.itercolumns()
    ):
        assert actual_column.type == expected_column.type
        assert actual_column.to_pylist() == expected_column.to_pylist()
        if pa.types.is_dictionary(actual_column.type):
            for output_chunk, expected_chunk in zip(
                actual_column.iterchunks(), expected_column.iterchunks()
            ):
                assert (
                    output_chunk.dictionary.to_pylist()
                    == expected_chunk.dictionary.to_pylist()
                )


@contextlib.contextmanager
def _render(params: Dict[str, Any], fetch_result: Optional[FetchResult]):
    with tempfile.NamedTemporaryFile() as empty_file:
        output_path = Path(empty_file.name)
        errors = render((), params, output_path, fetch_result=fetch_result)
        yield output_path, errors


def test_render_no_file():
    with _render(P(), None) as (path, errors):
        assert errors == []
        _assert_table_file(path, None)


def test_render_fetch_error():
    with tempfile.NamedTemporaryFile() as empty_file:
        fetch_errors = [RenderError(i18n_message("x", {"y": "z"}))]
        with _render(P(), FetchResult(Path(empty_file.name), fetch_errors)) as (
            path,
            errors,
        ):
            _assert_table_file(path, None)
            assert errors == [i18n_message("x", {"y": "z"})]


def test_render_deprecated_parquet():
    with _temp_parquet_file(pa.table({"A": [1, 2], "B": [3, 4]})) as fetched_path:
        with _render(P(), FetchResult(fetched_path)) as (path, errors):
            _assert_table_file(path, pa.table({"A": [1, 2], "B": [3, 4]}))
            assert errors == []


def test_render_deprecated_parquet_warning():
    with _temp_parquet_file(pa.table({"A": [1, 2], "B": [3, 4]})) as fetched_path:
        fetch_errors = [RenderError(i18n_message("truncated table"))]
        with _render(P(), FetchResult(fetched_path, fetch_errors)) as (path, errors):
            _assert_table_file(path, pa.table({"A": [1, 2], "B": [3, 4]}))
            assert errors == [i18n_message("truncated table")]


def test_render_deprecated_parquet_has_header_false():
    # Back in the day, we parsed during fetch. But has_header can change
    # between fetch and render. We were lazy, so we made fetch() follow the
    # most-common path: has_header=True. Then, in render(), we would "undo"
    # the change if has_header=False. This mangles the input data, but we
    # have no choice because we lost the input data. It was unwise. We have
    # abandoned supporting these files.
    with _temp_parquet_file(pa.table({"A": [1, 2], "B": [3, 4]})) as fetched_path:
        with _render(P(has_header=False), FetchResult(fetched_path)) as (path, errors):
            _assert_table_file(path, pa.table({"A": [1, 2], "B": [3, 4]}))
            assert errors == [i18n_message("error.parquet.cannotRemoveHeader")]


def test_render_has_header_true():
    with _temp_httpfile(
        "https://blah",
        "200 OK",
        b"A,B\na,b",
        headers=[("content-type", "text/csv")],
    ) as fetch_path:
        with _render(P(has_header=True), FetchResult(fetch_path)) as (path, errors):
            _assert_table_file(path, pa.table({"A": ["a"], "B": ["b"]}))
            assert errors == []


def test_render_has_header_false():
    with _temp_httpfile(
        "https://blah",
        "200 OK",
        b"1,2\n3,4",
        headers=[("content-type", "text/csv")],
    ) as fetch_path:
        with _render(P(has_header=False), FetchResult(fetch_path)) as (path, errors):
            _assert_table_file(
                path,
                pa.table(
                    {
                        "Column 1": pa.array([1, 3], pa.int8()),
                        "Column 2": pa.array([2, 4], pa.int8()),
                    }
                ),
            )
            assert errors == []


tests/test_migrate_params.py
from googlesheets import migrate_params


def test_v0_with_file():
    assert migrate_params(
        {
            "has_header": False,
            "version_select": "",
            "googlefileselect": (
                '{"id":"1AR-sdfsdf","name":"Filename","url":"https://docs.goo'
                'gle.com/a/org/spreadsheets/d/1MJsdfwer/view?usp=drive_web","'
                'mimeType":"text/csv"}'
            ),
        }
    ) == {
        "has_header": False,
        "version_select": "",
        "file": {
            "id": "1AR-sdfsdf",
            "name": "Filename",
            "url": (
                "https://docs.google.com/a/org/spreadsheets/"
                "d/1MJsdfwer/view?usp=drive_web"
            ),
            "mimeType": "text/csv",
        },
    }


def test_v0_no_file():
    assert migrate_params(
        {"has_header": False, "version_select": "", "googlefileselect": ""}
    ) == {"has_header": False, "version_select": "", "file": None}


def test_v1():
    assert migrate_params(
        {
            "has_header": False,
            "version_select": "",
            "file": {
                "id": "1AR-sdfsdf",
                "name": "Filename",
                "url": (
                    "https://docs.google.com/a/org/spreadsheets/"
                    "d/1MJsdfwer/view?usp=drive_web"
                ),
                "mimeType": "text/csv",
            },
        }
    ) == {
        "has_header": False,
        "version_select": "",
        "file": {
            "id": "1AR-sdfsdf",
            "name": "Filename",
            "url": (
                "https://docs.google.com/a/org/spreadsheets/"
                "d/1MJsdfwer/view?usp=drive_web"
            ),
            "mimeType": "text/csv",
        },
    }


googlesheets.py
import asyncio
import json
import os
from pathlib import Path
from typing import Any, Dict, List, NamedTuple, Optional

import cjwparquet
from cjwmodule.http import HttpError, httpfile
from cjwmodule.i18n import I18nMessage, trans
from cjwparse.api import MimeType, parse_file
from oauthlib import oauth2

GDRIVE_API_URL = "https://www.googleapis.com/drive/v3"  # unit tests override this


class RenderError(NamedTuple):
    """Mimics cjworkbench.cjwkernel.types.RenderError

    TODO move this to cjwmodule, so we can reuse it.
    """

    message: I18nMessage
    quick_fixes: List[None] = []


class FetchResult(NamedTuple):
    """Mimics cjworkbench.cjwkernel.types.FetchResult

    TODO move this to cjwmodule, so we can reuse it.
    """

    path: Path
    errors: List[RenderError] = []


def _generate_google_sheet_url(sheet_id: str) -> str:
    """
    URL to download text/csv from Google Drive.

    This uses the GDrive "export" API.

    Google Content-Type header is broken. According to RFC2616, "Data in
    character sets other than "ISO-8859-1" or its subsets MUST be labeled
    with an appropriate charset value". Google Sheets does not specify a
    charset (implying ISO-8859-1), but the text it serves is utf-8.

    So the caller should ignore the content-type Google returns.
    """
    return f"{GDRIVE_API_URL}/files/{sheet_id}/export?mimeType=text%2Fcsv"


def _generate_gdrive_file_url(sheet_id: str) -> str:
    """
    URL to download raw bytes from Google Drive.

    This discards Content-Type, including charset. GDrive doesn't know the
    charset anyway.
    """
    return f"{GDRIVE_API_URL}/files/{sheet_id}?alt=media"


def fetch_error(output_path: Path, message: I18nMessage):
    os.truncate(output_path, 0)
    return FetchResult(output_path, [RenderError(message)])


async def do_download(
    sheet_id: str, sheet_mime_type: str, oauth2_client: oauth2.Client, output_path: Path
) -> FetchResult:
    """
    Download spreadsheet from Google.

    If `sheet_mime_type` is 'application/vnd.google-apps.spreadsheet', use
    GDrive API to _export_ a text/csv. Otherwise, use GDrive API to _download_
    the file.
    """
    if sheet_mime_type == "application/vnd.google-apps.spreadsheet":
        url = _generate_google_sheet_url(sheet_id)
        sheet_mime_type = "text/csv"
    else:
        url = _generate_gdrive_file_url(sheet_id)
        # and use the passed sheet_mime_type

    url, headers, _ = oauth2_client.add_token(url, headers={})

    try:
        await httpfile.download(url, output_path, headers=headers)
    except HttpError.NotSuccess as err:
        response = err.response
        if response.status_code == 401:
            return fetch_error(
                output_path,
                trans(
                    "error.http.status401",
                    "Invalid credentials. Please reconnect to Google Drive.",
                ),
            )
        elif response.status_code == 403:
            return fetch_error(
                output_path,
                trans(
                    "error.http.status403",
                    "You chose a file your logged-in user cannot access. Please reconnect to Google Drive or choose a different file.",
                ),
            )
        elif response.status_code == 404:
            return fetch_error(
                output_path,
                trans(
                    "error.http.status404",
                    "File not found. Please choose a different file.",
                ),
            )
        else:
            return fetch_error(output_path, err.i18n_message)
    except HttpError as err:
        return fetch_error(output_path, err.i18n_message)

    return FetchResult(output_path)


def _render_deprecated_parquet(
    input_path: Path, errors: List[Any], output_path: Path, params: Dict[str, Any]
) -> List[I18nMessage]:
    cjwparquet.convert_parquet_file_to_arrow_file(input_path, output_path)
    if params["has_header"]:
        # In the deprecated parquet format, we _always_ parsed the header
        pass
    else:
        # We used to have a "moduleutils.turn_header_into_first_row()" but it
        # was broken by design (what about types???) and it was rarely used.
        # Let's not maintain it any longer.
        errors += [
            trans(
                "error.parquet.cannotRemoveHeader",
                "Please re-download this file to disable header-row handling",
            )
        ]

    return errors


def _calculate_mime_type(content_type: str) -> MimeType:
    for mime_type in MimeType:
        if content_type.startswith(mime_type.value):
            return mime_type
    # If we get here, we downloaded a MIME type we couldn't handle ... even
    # though we explicitly requested a MIME type we can handle. Undefined
    # behavior.
    raise RuntimeError("Unsupported content_type %s" % content_type)  # pragma: no cover


def _render_file(path: Path, params: Dict[str, Any], output_path: Path):
    with httpfile.read(path) as (parameters, status_line, headers, body_path):
        content_type = httpfile.extract_first_header(headers, "Content-Type")
        mime_type = _calculate_mime_type(content_type)
        # Ignore Google-reported charset. Google's headers imply latin-1 when
        # their data is utf-8.
        return parse_file(
            body_path,
            encoding=None,
            mime_type=mime_type,
            has_header=params["has_header"],
            output_path=output_path,
        )


def render(arrow_table, params, output_path, *, fetch_result, **kwargs):
    # Must perform header operation here in the event the header checkbox
    # state changes
    if fetch_result is None:
        # empty table
        return []
    elif fetch_result.path is not None and cjwparquet.file_has_parquet_magic_number(
        fetch_result.path
    ):
        # Deprecated files: we used to parse in fetch() and store the result
        # as Parquet. Now we've lost the original file data, and we need to
        # support our oldest users.
        #
        # In this deprecated format, parse errors were written as
        # fetch_result.errors.
        return _render_deprecated_parquet(
            fetch_result.path,
            [e.message for e in fetch_result.errors],
            output_path,
            params,
        )
    elif fetch_result.errors:
        # We've never stored errors+data. If there are errors, assume
        # there's no data.
        #
        # We've never stored errors with quick-fixes
        return [e.message for e in fetch_result.errors]
    else:
        assert not fetch_result.errors  # we've never stored errors+data.
        return _render_file(fetch_result.path, params, output_path)


def fetch_arrow(
    params: Dict[str, Any],
    secrets: Dict[str, Any],
    last_fetch_result: Any,
    input_table_parquet_path: Optional[Path],
    output_path: Path,
) -> FetchResult:
    file_meta = params["file"]
    if not file_meta:
        return fetch_error(
            output_path, trans("error.params.noFile", "Please choose a file")
        )

    # Ignore file_meta['url']. That's for the client's web browser, not for
    # an API request.
    sheet_id = file_meta["id"]
    if not sheet_id:
        # [adamhooper, 2019-12-06] has this ever happened?
        raise RuntimeError("Missing file.id")  # pragma: no cover

    # backwards-compat for old entries without 'mimeType', 2018-06-13
    sheet_mime_type = file_meta.get(
        "mimeType", "application/vnd.google-apps.spreadsheet"
    )

    secret = secrets.get("google_credentials")
    if not secret:
        return fetch_error(
            output_path,
            trans("error.secrets.noCredentials", "Please connect to Google Drive."),
        )
    if "error" in secret:
        return fetch_error(output_path, I18nMessage(**secret["error"]))
    assert "secret" in secret
    oauth2_client = oauth2.Client(
        client_id=None,  # unneeded
        token_type=secret["secret"]["token_type"],
        access_token=secret["secret"]["access_token"],
    )

    return asyncio.run(
        do_download(sheet_id, sheet_mime_type, oauth2_client, output_path)
    )


def _migrate_params_v0_to_v1(params):
    """
    v0: `googlefileselect` was a JSON-encoded String.

    v1: `file` is an Optional[Dict[str, str]]
    """
    if params["googlefileselect"]:
        file = json.loads(params["googlefileselect"])
    else:
        file = None
    return {
        "has_header": params["has_header"],
        "version_select": params["version_select"],
        "file": file,
    }


def migrate_params(params):
    if "googlefileselect" in params:
        params = _migrate_params_v0_to_v1(params)
    return params
